File: web/js/rle.js
// PackBits run-length encoding. Twin of internal/av/rle.go — the two must
// produce identical bytes for identical input (vector-enforced).
//
// Control byte c: [0,127] copy next c+1 literally; 128 no-op (tolerated on
// decode, never emitted); [129,255] repeat next byte 257-c times.

function runLen(src, i) {
  let j = i + 1;
  while (j < src.length && j - i < 128 && src[j] === src[i]) j++;
  return j - i;
}

export function packBits(src) {
  const dst = [];
  let i = 0;
  while (i < src.length) {
    const run = runLen(src, i);
    if (run >= 3) {
      dst.push(257 - run, src[i]);
      i += run;
      continue;
    }
    // Literal segment: extend until a 3+ run starts or 128 bytes.
    const start = i;
    i += run;
    while (i < src.length && i - start < 128) {
      const r = runLen(src, i);
      if (r >= 3) break;
      i += r;
    }
    if (i - start > 128) i = start + 128;
    dst.push(i - start - 1);
    for (let k = start; k < i; k++) dst.push(src[k]);
  }
  return Uint8Array.from(dst);
}

export function unpackBits(src, maxOut) {
  const dst = new Uint8Array(maxOut);
  let n = 0;
  let i = 0;
  while (i < src.length) {
    const c = src[i++];
    if (c < 128) {
      const len = c + 1;
      if (i + len > src.length) throw new Error("av: rle: truncated input");
      if (n + len > maxOut) throw new Error("av: rle: output exceeds limit");
      dst.set(src.subarray(i, i + len), n);
      n += len;
      i += len;
    } else if (c === 128) {
      // no-op
    } else {
      const len = 257 - c;
      if (i >= src.length) throw new Error("av: rle: truncated input");
      if (n + len > maxOut) throw new Error("av: rle: output exceeds limit");
      dst.fill(src[i], n, n + len);
      n += len;
      i++;
    }
  }
  return dst.subarray(0, n);
}

File: internal/av/rle.go
package av

import "errors"

// PackBits run-length encoding, the classic byte-oriented RLE (TIFF/MacPaint
// flavor). A control byte c means:
//
//	c in [0,127]:   copy the next c+1 bytes literally
//	c == 128:       no-op (never emitted by this encoder, tolerated on decode)
//	c in [129,255]: repeat the next byte 257-c times (a run of 2..128)
//
// It is chosen for legibility and because the payloads it sees are extreme:
// XOR deltas of temporally-stable dithered video are dominated by long zero
// runs, which cost 2 bytes per 128. This encoder emits runs only when they
// are 3+ bytes long; shorter repeats ride inside literals.

var (
	errRLETruncated = errors.New("av: rle: truncated input")
	errRLETooLarge  = errors.New("av: rle: output exceeds limit")
)

// runLen returns the length of the run of identical bytes at src[i:],
// capped at 128 (the longest a single PackBits token can express).
func runLen(src []byte, i int) int {
	j := i + 1
	for j < len(src) && j-i < 128 && src[j] == src[i] {
		j++
	}
	return j - i
}

// PackBits compresses src. The output is deterministic: the JS twin must
// produce identical bytes for identical input.
func PackBits(src []byte) []byte {
	dst := make([]byte, 0, len(src)/8+16)
	i := 0
	for i < len(src) {
		run := runLen(src, i)
		if run >= 3 {
			dst = append(dst, byte(257-run), src[i])
			i += run
			continue
		}
		// Literal segment: extend until a 3+ run starts or 128 bytes.
		start := i
		i += run
		for i < len(src) && i-start < 128 {
			r := runLen(src, i)
			if r >= 3 {
				break
			}
			i += r
		}
		if i-start > 128 {
			i = start + 128
		}
		dst = append(dst, byte(i-start-1))
		dst = append(dst, src[start:i]...)
	}
	return dst
}

// UnpackBits decompresses src, refusing to produce more than maxOut bytes
// (frames arrive from the network; the cap is the decoder's memory-safety
// boundary).
func UnpackBits(src []byte, maxOut int) ([]byte, error) {
	dst := make([]byte, 0, maxOut)
	i := 0
	for i < len(src) {
		c := src[i]
		i++
		switch {
		case c < 128: // literal of c+1 bytes
			n := int(c) + 1
			if i+n > len(src) {
				return nil, errRLETruncated
			}
			if len(dst)+n > maxOut {
				return nil, errRLETooLarge
			}
			dst = append(dst, src[i:i+n]...)
			i += n
		case c == 128: // no-op
		default: // run of 257-c copies
			n := 257 - int(c)
			if i >= len(src) {
				return nil, errRLETruncated
			}
			if len(dst)+n > maxOut {
				return nil, errRLETooLarge
			}
			for k := 0; k < n; k++ {
				dst = append(dst, src[i])
			}
			i++
		}
	}
	return dst, nil
}
